### PR TITLE
DM-31655: add AuxTel to ts_wep

### PR DESCRIPTION
**Pull request by suberlak**
This develops the capability to run Zernike estimation with ts_wep using auxTel data

### CODE REVIEW

**Comment by teweitsai** [comment]
This PR would break after the following PR is merged:
https://github.com/lsst-ts/ts_wep/pull/120

Please consider to wait until the above PR is merged, which had been approved. Thanks!

**Comment by teweitsai** [comment]
It looks like you are still working on the code. Please help to fix the failed Jenkins test. After you finished the update, please help to reorganize the `git` history and let me know to review the code. Thanks!

BTW, it looks like there is a lot of code change in this PR. Please consider to separate the update to multiple PRs to let me review. Otherwise, I could not promise any review quality for the huge amount of code change. Thanks!

**Comment by suberlak** [comment]
I cleaned the commit history.

**Comment by tribeiro** [comment]
I found the problem with the container that is causing the builds to fail.. Fix coming up soon..

**Comment by suberlak** [comment]
Thanks to Tiago's update Jenkins build now has passed and all tests pass.

File: python/lsst/ts/wep/Utility.py [hunk 1]
```diff
@@ -532,3 +532,62 @@ def writeCleanUpRepoCmd(repoDir, runName):
...
 
     return cleanUpCmd
+
+
+def getCamType(instName):
+    """ Get the camera type from instrument name.
+
+    Parameters
+    ----------
+    instName: str
```
Comment (teweitsai) [inline python/lsst/ts/wep/Utility.py]: instName: str --> instName : str

File: python/lsst/ts/wep/Utility.py [hunk 2]
```diff
@@ -532,3 +532,62 @@ def writeCleanUpRepoCmd(repoDir, runName):
...
+    if instName == 'lsst':
+        return CamType.LsstCam
+    elif instName == 'lsstfam':
+        return CamType.LsstFamCam
+    elif instName == "comcam":
+        return CamType.ComCam
+    elif instName == "auxTel":
+        return CamType.AuxTel
+    else:
+        raise ValueError("Instrument name (%s) is not supported." % instName)
```
Comment (teweitsai) [inline python/lsst/ts/wep/Utility.py]: Or:

```
raise ValueError(f"Instrument name ({instName}) is not supported.")
```

File: python/lsst/ts/wep/Utility.py [hunk 3]
```diff
@@ -532,3 +532,62 @@ def writeCleanUpRepoCmd(repoDir, runName):
...
+
+
+def getDefocalDisInMm(instName):
+    """
+    Get the defocal distance for the instrument
+
+    Parameters
+    ----------
+    instName: str
```
Comment (teweitsai) [inline python/lsst/ts/wep/Utility.py]: instName: str --> instName : str

File: python/lsst/ts/wep/Utility.py [hunk 4]
```diff
@@ -532,3 +532,62 @@ def writeCleanUpRepoCmd(repoDir, runName):
...
+        Instrument name, one of
+        'lsst', 'lsstfam', 'comcam',
+        'auxTel'
+
+    Returns
+    -------
+    defocalDisInMm: float
```
Comment (teweitsai) [inline python/lsst/ts/wep/Utility.py]: defocalDisInMm: float --> defocalDisInMm : float

File: python/lsst/ts/wep/Utility.py [hunk 5]
```diff
@@ -532,3 +532,62 @@ def writeCleanUpRepoCmd(repoDir, runName):
...
+        Defocal distance in mm
```
Comment (teweitsai) [inline python/lsst/ts/wep/Utility.py]: Period in the end.

File: python/lsst/ts/wep/Utility.py [hunk 6]
```diff
@@ -532,3 +532,62 @@ def writeCleanUpRepoCmd(repoDir, runName):
...
+    ------
+    ValueError
+        Instrument name is not supported.
+    """
+    if instName in ['lsst', 'lsstfam', 'comcam']:
+        return 1.5
+    elif instName == "auxTel":
+        return 0.8
+    else:
+        raise ValueError("Instrument name (%s) is not supported." % instName)
```
Comment (teweitsai) [inline python/lsst/ts/wep/Utility.py]: Or:

```
raise ValueError(f"Instrument name ({instName}) is not supported.")
```

File: tests/test_utility.py [hunk 7]
```diff
@@ -225,6 +228,18 @@ def testWriteCleanUpRepoCmd(self):
...
+        self.assertEqual(getDefocalDisInMm('lsstfam'), 1.5)
+        self.assertEqual(getDefocalDisInMm('comcam'), 1.5)
+        self.assertEqual(getDefocalDisInMm('auxTel'), 0.8)
+
+    def testGetCamType(self):
+        self.assertEqual(getCamType('lsst'), CamType.LsstCam)
+        self.assertEqual(getCamType('lsstfam'), CamType.LsstFamCam)
+        self.assertEqual(getCamType('comcam'), CamType.ComCam)
+        self.assertEqual(getCamType('auxTel'), CamType.AuxTel)
+
```
Comment (teweitsai) [inline tests/test_utility.py]: Please help to test the raised `ValueError` as well.

File: tests/testData/pipelineConfigs/testLatissPipeline.yaml [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      opticalModel: 'onAxis'
+  estimateZernikesLatissTask:
+    class: lsst.ts.wep.task.EstimateZernikesLatissTask.EstimateZernikesLatissTask
+    config:
+      # And here we specify the configuration settings
+      donutTemplateSize: 200
+      donutStampSize: 200
+      initialCutoutPadding: 40
+      instName: 'auxTel'
+      opticalModel: 'onAxis'
```
Comment (teweitsai) [inline tests/testData/pipelineConfigs/testLatissPipeline.yaml]: One new line in the end.

File: tests/testData/pipelineConfigs/testDonutDirectDetectPipeline.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      doDark: False
+      doFlat: False
+      doApplyGains: True
+      doFringe: False
+      doOverscan: True
+  generateDonutDirectDetectTask:
+    class: lsst.ts.wep.task.GenerateDonutDirectDetectTask.GenerateDonutDirectDetectTask
+    config:
+      donutTemplateSize: 160
+      removeBlends: True
```
Comment (teweitsai) [inline tests/testData/pipelineConfigs/testDonutDirectDetectPipeline.yaml]: One new line in the end.

File: python/lsst/ts/wep/DonutDetector.py [hunk 10]
```diff
@@ -71,7 +80,19 @@ def detectDonuts(
...
+
+        elif binaryChoice == 'deblend':
+            deblend = DeblendAdapt()
+            binaryExp = deblend._getImgBinaryAdapt(copy(expArray))
+
+        elif binaryChoice == 'exposure':
+            binaryExp = copy(expArray)
+
+        else:
+            raise ValueError(f"binaryChoice {binaryChoice} is not supported.")
```
Comment (teweitsai) [inline python/lsst/ts/wep/DonutDetector.py:99]: Please help to add the `ValueError` to the function doc string.

File: python/lsst/ts/wep/cwfs/DonutTemplateModel.py [hunk 11]
```diff
@@ -65,34 +67,44 @@ def makeTemplate(
...
+            sensorName = list(camera.getNameIter())[0]  # only one detector in latiss
+            detector = camera.get(sensorName)
+            xp, yp = detector.getCenter(cameraGeom.FOCAL_PLANE)  # center of CCD in mm
+
+            # multiply by 1000 to for mm --> microns conversion
+            sensorXMicron = yp * 1000
+            sensorYMicron = xp * 1000
+
+        else:
+            raise ValueError("Camera type (%s) is not supported." % camType)
```
Comment (teweitsai) [inline python/lsst/ts/wep/cwfs/DonutTemplateModel.py:113]: Please help to document the ValueError in function doc string.

File: python/lsst/ts/wep/cwfs/DonutTemplateModel.py [hunk 12]
```diff
@@ -65,34 +67,44 @@ def makeTemplate(
...
+
+            sensorXMicron, sensorYMicron = np.array(
+                focalPlaneLayout[sensorName][:2], dtype=float
+            )
+
+        elif camType == CamType.AuxTel:
+            # Defocal distance for Latiss in mm
+            # for LsstCam can use the default
+            # hence only need to set here
+            announcedDefocalDisInMm = 0.8
```
Comment (teweitsai) [inline python/lsst/ts/wep/cwfs/DonutTemplateModel.py]: How about to use `getDefocalDisInMm()` to get this `0.8` value to minimize the use of magic number?

File: python/lsst/ts/wep/task/EstimateZernikesBase.py [hunk 13]
```diff
@@ -163,6 +179,15 @@ def getTemplate(self, detectorName, defocalType):
             Name of the CCD (e.g. 'R22_S11').
         defocalType: enum 'DefocalType'
             Defocal type of the donut image.
+        donutTemplateSize: int
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesBase.py]: donutTemplateSize: int --> donutTemplateSize : int

In addition, the above doc string of `detectorName` and `defocalType` as well.

File: python/lsst/ts/wep/task/EstimateZernikesLatissTask.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+            focuszPair = (focusZVal0, focusZVal1)
+            extraExp = min(focuszPair)
+            intraExp = max(focuszPair)
+
+            expDicIdx = {focusZVal0: 0, focusZVal1: 1}
+            intraExpIdx = expDicIdx[intraExp]
+            extraExpIdx = expDicIdx[extraExp]
+        else:
+            # Need two different focusz values
+            raise ValueError(errorStr)
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesLatissTask.py]: Maybe:

```
raise ValueError("Must have two images with different FOCUSZ parameter.")
```

and remove the above `errorStr`.

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        ----------
+        donutCat : pandas.DataFrame
+            The donut catalog from running DonutDetector,
+            contains columns 'y_center', 'x_center'
+        exposure : lsst.afw.image.Exposure
+            Exposure with the donut images.
+
+        Returns
+        -------
+        donutCat: pandas.DataFrame
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py]: donutCat: pandas.DataFrame --> donutCat : pandas.DataFrame

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+            Donut catalog with reorganized content.
+        """
+        # EstimateZernikes  expects the following column names:
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py]: There are two spaces between "EstimateZernikes" and "expects".

File: python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        templateMaker = DonutTemplateFactory.createDonutTemplate(
+            DonutTemplateType.Model
+        )
+
+        template = templateMaker.makeTemplate(
+            detectorName, defocalType, self.donutTemplateSize,
+            camType, self.opticalModel, pixelScale
+        )
+
+        # given this template, detect donuts  in one of the defocal images
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/GenerateDonutDirectDetectTask.py]: There are two spaces between "donuts" and "in".

File: tests/task/test_generateDonutDirectDetectTask.py [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+            "detector": 94,
+            "exposure": 4021123106001,
+        }
+        testExposure = self.butler.get(
+            "raw", dataId=testDataId, collections="LSSTCam/raw/all"
+        )
+
+        # setup a test donut catalog DataFrame
+        x_center = np.arange(15)
+        y_center = 0.5*np.arange(15)
```
Comment (teweitsai) [inline tests/task/test_generateDonutDirectDetectTask.py]: I thought `black` will do `y_center = 0.5 * np.arange(15)`?

**Review by teweitsai — APPROVED** [review]
Looks good. Please help to update the following documents as well:
1. doc/uml/taskClass.uml
2. doc/content.rst

Thanks!

File: python/lsst/ts/wep/Utility.py (hunk 1 above)
Comment (suberlak) [reply python/lsst/ts/wep/Utility.py]: I'm not sure I understand- are any changes implied here ?

Comment (teweitsai) [reply python/lsst/ts/wep/Utility.py]: I mean, a space between "instName" and ":".

**Pull request merged**